File: apps/robot/src/robot.ts
import {
  boundCommand,
  DEFAULT_SPEED,
  type CommandMessage,
  type FaceState,
  type TelemetryMessage,
} from "./protocol.ts";
import type { Transport } from "./transport.ts";

export interface RobotOptions {
  /** How often to send a heartbeat while connected. The device agent stops the motors and
   *  waits for a fresh one if this goes silent (apps/robot/PLAN.md, Phase 1 responsibilities) —
   *  a dropped Wi-Fi link must not leave the robot driving blind. */
  heartbeatIntervalMs?: number;
  /** How long a command may go unacknowledged before Robot gives up on it. */
  commandTimeoutMs?: number;
}

const DEFAULT_HEARTBEAT_INTERVAL_MS = 500;
const DEFAULT_COMMAND_TIMEOUT_MS = 3000;

export class RobotCommandError extends Error {}
export class RobotTimeoutError extends Error {}

interface PendingRequest {
  resolve: (message: TelemetryMessage) => void;
  reject: (error: Error) => void;
  timeout: ReturnType<typeof setTimeout>;
}

/**
 * Laptop-side robot API (apps/robot/PLAN.md, Phase 2). This is the only thing an LLM tool call
 * or any other Rocky code should touch — never the transport or protocol modules directly —
 * because every command here is bounded (protocol.ts's boundCommand) before it reaches the wire.
 */
export class Robot {
  private nextId = 0;
  private pending = new Map<string, PendingRequest>();
  private heartbeatTimer: ReturnType<typeof setInterval> | undefined;
  private statusListeners: Array<(status: { battery: number; connected: boolean }) => void> = [];

  constructor(
    private readonly transport: Transport,
    private readonly options: RobotOptions = {},
  ) {
    this.transport.onTelemetry((message) => this.handleTelemetry(message));
  }

  async connect(): Promise<void> {
    await this.transport.connect();
    const intervalMs = this.options.heartbeatIntervalMs ?? DEFAULT_HEARTBEAT_INTERVAL_MS;
    this.heartbeatTimer = setInterval(() => {
      this.transport.send({ id: this.allocateId(), type: "heartbeat" });
    }, intervalMs);
  }

  disconnect(): void {
    if (this.heartbeatTimer) clearInterval(this.heartbeatTimer);
    for (const [id, request] of this.pending) {
      clearTimeout(request.timeout);
      request.reject(new RobotTimeoutError(`disconnected while "${id}" was in flight`));
    }
    this.pending.clear();
    this.transport.close();
  }

  async drive(distanceCm: number, speed: number = DEFAULT_SPEED): Promise<void> {
    await this.sendCommand({ id: this.allocateId(), type: "drive", distanceCm, speed });
  }

  async turn(degrees: number, speed: number = DEFAULT_SPEED): Promise<void> {
    await this.sendCommand({ id: this.allocateId(), type: "turn", degrees, speed });
  }

  async stop(): Promise<void> {
    await this.sendCommand({ id: this.allocateId(), type: "stop" });
  }

  async setFace(face: FaceState): Promise<void> {
    await this.sendCommand({ id: this.allocateId(), type: "setFace", face });
  }

  async setLights(r: number, g: number, b: number): Promise<void> {
    await this.sendCommand({ id: this.allocateId(), type: "setLights", r, g, b });
  }

  async readDistance(): Promise<number> {
    const reply = await this.sendCommand({ id: this.allocateId(), type: "readDistance" });
    if (reply.type !== "distance") {
      throw new RobotCommandError(`expected a distance reply, got "${reply.type}"`);
    }
    return reply.cm;
  }

  async readLineSensors(): Promise<number[]> {
    const reply = await this.sendCommand({ id: this.allocateId(), type: "readLineSensors" });
    if (reply.type !== "lineSensors") {
      throw new RobotCommandError(`expected a lineSensors reply, got "${reply.type}"`);
    }
    return reply.values;
  }

  /** Unprompted battery/connectivity beacons the agent sends outside the request/response flow. */
  onStatus(listener: (status: { battery: number; connected: boolean }) => void): void {
    this.statusListeners.push(listener);
  }

  private allocateId(): string {
    this.nextId += 1;
    return String(this.nextId);
  }

  private sendCommand(command: CommandMessage): Promise<TelemetryMessage> {
    const bounded = boundCommand(command);
    return new Promise((resolve, reject) => {
      const timeoutMs = this.options.commandTimeoutMs ?? DEFAULT_COMMAND_TIMEOUT_MS;
      const timeout = setTimeout(() => {
        this.pending.delete(bounded.id);
        reject(new RobotTimeoutError(`"${bounded.type}" (${bounded.id}) timed out`));
      }, timeoutMs);
      this.pending.set(bounded.id, { resolve, reject, timeout });
      this.transport.send(bounded);
    });
  }

  private handleTelemetry(message: TelemetryMessage): void {
    if (message.type === "status") {
      for (const listener of this.statusListeners) listener(message);
      return;
    }
    const request = this.pending.get(message.id);
    if (!request) return;
    this.pending.delete(message.id);
    clearTimeout(request.timeout);
    if (message.type === "error") {
      request.reject(new RobotCommandError(message.message));
    } else {
      request.resolve(message);
    }
  }
}

File: apps/ios/Rocky/Sources/Robot.swift
import Foundation

/// The only thing app code should touch -- never RobotTCPTransport or the protocol types
/// directly -- because every command here is bounded (boundCommand) before it reaches the wire.
/// An actor, not a class-plus-DispatchQueue: serializes pending-request/heartbeat state without
/// the manual-sync deadlock risk a shared GCD queue would carry once callbacks start calling
/// back into it. Mirrors apps/robot/src/robot.ts.
actor Robot {
    private let transport: RobotTCPTransport
    private let heartbeatInterval: TimeInterval
    private let commandTimeout: TimeInterval

    private var nextId = 0
    private var pending: [String: CheckedContinuation<TelemetryMessage, Error>] = [:]
    private var heartbeatTask: Task<Void, Never>?

    var onStatus: (@Sendable (_ battery: Double, _ connected: Bool) -> Void)?
    var onDisconnect: (@Sendable (Error?) -> Void)?
    /// A drive or turn physically began (the board's `started` reply). Carries the command id.
    /// This is what turns an assumed action into a confirmed one -- see
    /// apps/ios/docs/embodiment.md.
    var onStarted: (@Sendable (_ id: String) -> Void)?
    /// Anything at all arrived from the board, including heartbeat replies. Positive evidence
    /// that the interpreter is alive, which an open socket is not on this hardware.
    var onTraffic: (@Sendable () -> Void)?

    /// Actor-isolated properties can't be assigned from outside, so callers install their
    /// handlers through this rather than through `robot.onStarted = ...`.
    func observe(
        started: (@Sendable (_ id: String) -> Void)? = nil,
        traffic: (@Sendable () -> Void)? = nil,
        disconnected: (@Sendable (Error?) -> Void)? = nil
    ) {
        if let started { onStarted = started }
        if let traffic { onTraffic = traffic }
        if let disconnected { onDisconnect = disconnected }
    }

    init(
        host: String,
        port: UInt16 = 8765,
        heartbeatInterval: TimeInterval = 0.5,
        commandTimeout: TimeInterval = 3.0
    ) {
        self.transport = RobotTCPTransport(host: host, port: port)
        self.heartbeatInterval = heartbeatInterval
        self.commandTimeout = commandTimeout
    }

    func connect() async throws {
        transport.onMessage = { [weak self] message in
            Task { await self?.handleTelemetry(message) }
        }
        transport.onDisconnect = { [weak self] error in
            Task { await self?.handleDisconnect(error) }
        }
        try await transport.connect()
        startHeartbeat()
    }

    func disconnect() {
        heartbeatTask?.cancel()
        heartbeatTask = nil
        for continuation in pending.values {
            continuation.resume(throwing: RobotError.disconnected)
        }
        pending.removeAll()
        transport.close()
    }

    /// `id` and `timeout` are explicit because a drive is only acked when it *finishes*, so the
    /// caller -- which knows the requested distance and speed -- is the only thing that can say
    /// how long finishing ought to take. A fixed timeout here used to report a perfectly healthy
    /// six-second drive as a failure three seconds in.
    @discardableResult
    func drive(
        id: String,
        distanceCm: Double,
        speed: Double = RobotLimits.defaultSpeed,
        timeout: TimeInterval? = nil
    ) async throws -> TelemetryMessage {
        try await send(.drive(id: id, distanceCm: distanceCm, speed: speed), timeout: timeout)
    }

    @discardableResult
    func turn(
        id: String,
        degrees: Double,
        speed: Double = RobotLimits.defaultSpeed,
        timeout: TimeInterval? = nil
    ) async throws -> TelemetryMessage {
        try await send(.turn(id: id, degrees: degrees, speed: speed), timeout: timeout)
    }

    /// The next wire id, so a caller can correlate `started`/`ack` replies with its own action
    /// before the command has even been sent.
    func nextCommandId() -> String { allocateId() }

    @discardableResult
    func stop() async throws -> TelemetryMessage {
        try await send(.stop(id: allocateId()))
    }

    func setFace(_ face: FaceState) async throws {
        _ = try await send(.setFace(id: allocateId(), face: face))
    }

    func setLights(r: Int, g: Int, b: Int) async throws {
        _ = try await send(.setLights(id: allocateId(), r: r, g: g, b: b))
    }

    func readDistance() async throws -> Double {
        let reply = try await send(.readDistance(id: allocateId()))
        guard case .distance(_, let cm) = reply else {
            throw RobotError.unexpectedReply("expected a distance reply, got \(reply)")
        }
        return cm
    }

    // MARK: - Private

    private func allocateId() -> String {
        nextId += 1
        return String(nextId)
    }

    private func send(_ command: CommandMessage, timeout: TimeInterval? = nil) async throws -> TelemetryMessage {
        let bounded = boundCommand(command)
        let deadline = timeout ?? commandTimeout
        return try await withCheckedThrowingContinuation { continuation in
            pending[bounded.id] = continuation
            transport.send(bounded)
            let id = bounded.id
            let type = bounded.type
            Task {
                try? await Task.sleep(nanoseconds: UInt64(deadline * 1_000_000_000))
                await self.failIfStillPending(id, type: type)
            }
        }
    }

    private func failIfStillPending(_ id: String, type: String) {
        if let continuation = pending.removeValue(forKey: id) {
            continuation.resume(throwing: RobotError.timedOut("\"\(type)\" (\(id))"))
        }
    }

    private func handleTelemetry(_ message: TelemetryMessage) {
        onTraffic?()
        if case .status(let battery, let connected) = message {
            onStatus?(battery, connected)
            return
        }
        // Neither of these ends the command they refer to. `started` says the maneuver began and
        // the ack is still to come; `pong` refers to a heartbeat that has no continuation waiting
        // on it at all. Resuming on either would finish a drive the instant it began.
        if case .started(let id) = message {
            onStarted?(id)
            return
        }
        if case .pong = message { return }
        guard let id = message.id, let continuation = pending.removeValue(forKey: id) else { return }
        if case .error(_, let errorMessage) = message {
            continuation.resume(throwing: RobotError.commandFailed(errorMessage))
        } else {
            continuation.resume(returning: message)
        }
    }

    private func handleDisconnect(_ error: Error?) {
        heartbeatTask?.cancel()
        heartbeatTask = nil
        for continuation in pending.values {
            continuation.resume(throwing: error ?? RobotError.disconnected)
        }
        pending.removeAll()
        onDisconnect?(error)
    }

    private func startHeartbeat() {
        heartbeatTask?.cancel()
        heartbeatTask = Task { [weak self] in
            await self?.runHeartbeatLoop()
        }
    }

    private func runHeartbeatLoop() async {
        while !Task.isCancelled {
            transport.send(.heartbeat(id: allocateId()))
            try? await Task.sleep(nanoseconds: UInt64(heartbeatInterval * 1_000_000_000))
        }
    }
}
